hypothesis/lms
# Emit pyramid events to store them in the DB

For #3947 

The previous PRs this depends on are not yet merged.

https://github.com/hypothesis/lms/pull/4282
https://github.com/hypothesis/lms/pull/4283


Do not merge yet.



This should allow us to keep track of launches in LTI1.3 while also making the table flexible enough to track other events.


## Testing 

- Start with a fresh DB and in `main` branch

```git checkout main; docker stop lms_postgres_1; docker rm lms_postgres_1; make services db devdata```


- Launch a few assignment (still on main):

https://hypothesis.instructure.com/courses/125/assignments/873
https://hypothesis.instructure.com/courses/125/assignments/1513

and from another course

https://hypothesis.instructure.com/courses/252/assignments/1323



- Check the launches table

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "select * from lti_launches"

 id |          created           |                context_id                |                  lti_key                   
----+----------------------------+------------------------------------------+--------------------------------------------
  6 | 2022-08-03 15:56:59.875628 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  7 | 2022-08-03 15:57:04.053202 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  8 | 2022-08-03 15:57:05.850989 | 3216ae11f12d910d3e8feb12d1b2e2a219185a35 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
(3 rows)
```


- Switch to this branch and apply the migrations

```
git checkout lti-events; hdev alembic upgrade head
INFO  [alembic.runtime.migration] Running upgrade 1e146996cca6 -> 42b39684836f, Add events, event_type, event_user and event_data tables.
INFO  [alembic.runtime.migration] Running upgrade 42b39684836f -> 48955a83f040, Backfills the events table using the data from  `lti_launches`.
```

### Check the data migration

- Check that `lti_lauches` is now a view

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "\dv lti_launches"
            List of relations
 Schema |     Name     | Type |  Owner   
--------+--------------+------+----------
 public | lti_launches | view | postgres
(1 row)
```


- And the data should be accessible there:


```
tox -qe dockercompose -- exec postgres psql -U postgres -c "select * from lti_launches"

 id |          created           |                context_id                |                  lti_key                   
----+----------------------------+------------------------------------------+--------------------------------------------
  2 | 2022-08-03 15:57:04.053202 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  1 | 2022-08-03 15:56:59.875628 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  3 | 2022-08-03 15:57:05.850989 | 3216ae11f12d910d3e8feb12d1b2e2a219185a35 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
(3 rows)
```

The IDs would be different (now from the event table). The rest should match.



### Check the new table functionality

- Launch a few assignments again

https://hypothesis.instructure.com/courses/125/assignments/873
https://hypothesis.instructure.com/courses/125/assignments/1513
https://hypothesis.instructure.com/courses/252/assignments/1323

- Also configure an assignment, for example 

https://hypothesis.instructure.com/courses/125/assignments/3118/edit?name=Testing+upgrade+-+marcos&due_at=null&points_possible=10


- Query the new events tables

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "select event.*, user_id, lti_role_id, extra from event left join event_user on event_id = event.id left join event_data on event_data.event_id = event.id"
 id |         timestamp          | type_id | application_instance_id | course_id | assignment_id | grouping_id | user_id | lti_role_id |                                        extra                                         
----+----------------------------+---------+-------------------------+-----------+---------------+-------------+---------+-------------+--------------------------------------------------------------------------------------
  4 | 2022-08-03 15:59:46.364437 |       1 |                       8 |         1 |            25 |             |       1 |           1 | 
  5 | 2022-08-03 15:59:50.792776 |       1 |                       8 |         1 |            26 |             |       1 |           1 | 
  6 | 2022-08-03 15:59:53.647374 |       1 |                       8 |         3 |            27 |             |       1 |           1 | 
  7 | 2022-08-03 16:00:13.291383 |       2 |                       8 |         1 |               |             |       1 |           1 | {"document_url": "http://localhost:8001/lti_launches?url=https%3A%2F%2Fexample.com"}
  2 | 2022-08-03 15:57:04.053202 |       1 |                       8 |         1 |               |             |         |             | 
  1 | 2022-08-03 15:56:59.875628 |       1 |                       8 |         1 |               |             |         |             | 
  3 | 2022-08-03 15:57:05.850989 |       1 |                       8 |         3 |               |             |         |             | 
(7 rows)
```

- Note the three rows migrated from `lti_launches` before, those don't have user or assignment
- The new launches do have user and assignment
- There's also a row for the deep linking event. No assignment but extra data.



- Check the lti_launches view 

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "select * from lti_launches"

id |          created           |                context_id                |                  lti_key                   
----+----------------------------+------------------------------------------+--------------------------------------------
  6 | 2022-08-03 15:59:53.647374 | 3216ae11f12d910d3e8feb12d1b2e2a219185a35 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  3 | 2022-08-03 15:57:05.850989 | 3216ae11f12d910d3e8feb12d1b2e2a219185a35 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  5 | 2022-08-03 15:59:50.792776 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  4 | 2022-08-03 15:59:46.364437 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  2 | 2022-08-03 15:57:04.053202 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
  1 | 2022-08-03 15:56:59.875628 | f3cd019017839c4630358662a05540f2f6ec5f93 | Hypothesisb3be0b33d0b5e4b1cf3aaf04e0e1819a
(6 rows)
```

The row for deep linking is not present. But the new ones, created after the migration are here.
